### PR TITLE
add sonar action version 2.0

### PR DESCRIPTION
**Pull request by chengjoey**
## Description

[erda issue link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=287301&iterationID=1115&pId=0&type=TASK)

## Checklist

 - [ ] I made sure to check the compatibility of the erda version statemented in the action's spec.yml.
 - [ ] My change is adequately tested.

### CODE REVIEW

**Comment by Effet** [comment]
how to compatible with 1.0?

**Comment by chengjoey** [comment]
> how to compatible with 1.0?

image
set 1.0 as default

File: actions/sonar/2.0/internal/web_api.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		metricSort = MetricKeyUncoveredLines.String()
+	case MeasureTypeDuplications:
+		metricKeys = strings.Join([]string{
+			MetricKeyDuplicatedLinesDensity.String(),
+			MetricKeyDuplicatedLines.String(),
+		}, ",")
+		metricSort = MetricKeyDuplicatedLinesDensity.String()
+	default:
+		return nil, errors.Errorf("error type, key: %s", metricKeys)
+	}
```
Comment (Effet) [inline actions/sonar/2.0/internal/web_api.go:50]: what these code use for?

**Review by Effet — CHANGES_REQUESTED** [review]

File: actions/sonar/2.0/internal/web_api.go (hunk 1 above)
Comment (chengjoey) [reply actions/sonar/2.0/internal/web_api.go]: adjuested, removed useless code

Comment (chengjoey) [reply actions/sonar/2.0/internal/web_api.go]: image

File: actions/sonar/2.0/internal/analysis.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+			continue
+		}
+		results.Add(ResultKey(fmt.Sprintf("metric: %s", condition.MetricKey)), string(condition.Status))
+		results.Add(ResultKey(fmt.Sprintf("metric(detail): %s", condition.MetricKey)), jsonparse.JsonOneLine(condition))
+	}
+
+	if cfg.ActionParams.MustGateStatusOK && qualityGateStatus != SonarQualityGateStatusOK {
+		logrus.Errorf("QUALITY GATE STATUS: %s", qualityGateStatus)
+		return results, fmt.Errorf("sonar quality gate status is not ok, status: %s", qualityGateStatus)
+	}
```
Comment (Effet) [inline actions/sonar/2.0/internal/analysis.go:99]: I suggest "one func do one thing", and `Analysis` should just return the analysis result, you can move this gate check logic to outside.

**Review by Effet — CHANGES_REQUESTED** [review]

File: actions/sonar/2.0/internal/application.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/pkg/http/httpclient"
+)
+
+func getApplication(cfg *Conf) (*apistructs.ApplicationDTO, error) {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply actions/sonar/2.0/internal/application.go:11]: non

File: actions/sonar/2.0/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+### Sonar Action 2.0
+
+使用 sonar 进行代码质量检测
+
+#### 使用
+
```
Comment (Effet) [inline actions/sonar/2.0/README.md:6]: give another simple example which sonar_host_url in application setting page

File: actions/sonar/2.0/internal/pkg/conf.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	// SonarHostURL sonar 服务器地址，用户可以手动指定。若不填写，则使用平台提供的 sonar 服务
+	SonarHostURL string `env:"ACTION_SONAR_HOST_URL"`
+	// SonarLogin is the login or authentication token of a SonarQube user with Execute Analysis permission on the project
+	SonarLogin string `env:"ACTION_SONAR_LOGIN"`
+	// SonarPassword is the password that goes with the sonar.login username. This should be left blank if an authentication token is being used
+	SonarPassword string `env:"ACTION_SONAR_PASSWORD"`
+
+	// Project
+	// +optional
+	ProjectKey string `env:"ACTION_PROJECT_KEY"`
```
Comment (Effet) [inline actions/sonar/2.0/internal/pkg/conf.go:19]: `ACTION_SONAR_PROJECT_KEY` maybe better

File: actions/sonar/2.0/internal/pkg/sonar.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	SonarQualityGateStatusWARN  SonarQualityGateStatus = "WARN"
+	SonarQualityGateStatusERROR SonarQualityGateStatus = "ERROR"
+	SonarQualityGateStatusNONE  SonarQualityGateStatus = "NONE"
+)
+
+// MetricKey 指标名
+// see: https://docs.sonarqube.org/latest/user-guide/metric-definitions/
+type MetricKey string
+
+var (
```
Comment (Effet) [inline actions/sonar/2.0/internal/pkg/sonar.go:106]: is there any sonar go sdk to use?

File: pkg/envconf/platform.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package envconf
+
+type PlatformParams struct {
+	Debug bool `env:"ACTION_DEBUG" default:"false"`
```
Comment (Effet) [inline pkg/envconf/platform.go:4]: it looks like not a platform param?

**Review by Effet — CHANGES_REQUESTED** [review]

File: actions/sonar/2.0/README.md (hunk 4 above)
Comment (chengjoey) [reply actions/sonar/2.0/README.md:6]: maybe add example when `erda-cli` can be used

File: actions/sonar/2.0/internal/pkg/conf.go (hunk 5 above)
Comment (chengjoey) [reply actions/sonar/2.0/internal/pkg/conf.go]: adjusted

File: actions/sonar/2.0/internal/pkg/sonar.go (hunk 6 above)
Comment (chengjoey) [reply actions/sonar/2.0/internal/pkg/sonar.go:106]: not found golang sonar sdk

File: pkg/envconf/platform.go (hunk 7 above)
Comment (chengjoey) [reply pkg/envconf/platform.go]: adjusted, move this to action param

File: pkg/meta/result_meta.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+func (r *ResultMetas) Add(key string, value string) {
+	*r = append(*r, ResultMeta{Key: key, Value: value})
+}
+
+func NewResultMetas() ResultMetas {
+	return make(ResultMetas, 0)
+}
+
+func (r *ResultMetas) Store(metaFilepath string) error {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply pkg/meta/result_meta.go]: impl `logCollector` and `fileCollector` to collect result meta, default use `action meta:k=v`
image

File: pkg/envconf/platform.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	AppID        uint64 `env:"DICE_APPLICATION_ID"`
+	AppName      string `env:"DICE_APPLICATION_NAME"`
+	Workspace    string `env:"DICE_WORKSPACE"`
+	GittarRepo   string `env:"GITTAR_REPO"`
+	GittarBranch string `env:"GITTAR_BRANCH"`
+	GittarCommit string `env:"GITTAR_COMMIT"`
+	OperatorID   string `env:"DICE_OPERATOR_ID"`
+	PipelineID   int64  `env:"PIPELINE_ID"`
+
+	OrgId  uint64 `env:"DICE_ORG_ID"`
```
Comment (Effet) [inline pkg/envconf/platform.go:15]: it should be `OrgID`

File: actions/sonar/2.0/internal/pkg/sonar.go [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+	"time"
+
+	"github.com/erda-project/erda-actions/pkg/command"
+	"github.com/erda-project/erda-actions/pkg/meta"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda-actions/pkg/dice"
+	"github.com/erda-project/erda/pkg/encoding/jsonparse"
+	"github.com/erda-project/erda/pkg/http/httpclient"
```
Comment (Effet) [inline actions/sonar/2.0/internal/pkg/sonar.go:19]: please format the import order

**Review by Effet — CHANGES_REQUESTED** [review]

**Review by Effet — APPROVED** [review]

**Pull request merged**